src/charm.py

```diff
@@ -194,30 +196,47 @@ def _update_mimir_config(self) -> bool:
         return False
 
     def _set_data_dirs(self, config: Dict[str, Any]) -> dict:
-        """Set the data-dirs in the config we received from the coordinator.
+        """Set the data directories in the received config from the coordinator.
 
-        - Place all data dirs under a common root data dir, so files are persisted across upgrades.
-          Following the default names from official docs:
-          https://grafana.com/docs/mimir/latest/references/configuration-parameters/
+        - All data directories are placed under a common root data directory to persist
+        files across upgrades. The naming follows the default conventions from the
+        official Mimir docs: https://grafana.com/docs/mimir/latest/references/configuration-parameters/
         """
         config = config.copy()
-        for key, folder in (
-            ("alertmanager", "data-alertmanager"),
-            ("compactor", "data-compactor"),
-        ):
-            if key not in config:
-                config[key] = {}
-            config[key]["data_dir"] = str(self._root_data_dir / folder)
-
-        # blocks_storage:
-        #   bucket_store:
-        #     sync_dir: /etc/mimir/tsdb-sync
-        #   data_dir: /data/tsdb-sync
-        if config.get("blocks_storage"):
-            config["blocks_storage"] = {
-                "bucket_store": {"sync_dir": str(self._root_data_dir / "tsdb-sync")}
-            }
 
+        # Define a list of keys, subkeys, and folders in the Mimir config
+        # that need to be under the common root data directory
+        data_mapping = [
+            ("alertmanager", "data_dir", "data-alertmanager"),
+            ("alertmanager_storage", "filesystem", "data-alertmanager-recovery"),
+            ("compactor", "data_dir", "data-compactor"),
+            ("ruler", "rule_path", "data-ruler"),
+            ("ruler_storage", "filesystem", "ruler"),
+            ("blocks_storage", "filesystem", "blocks"),
+            ("blocks_storage", "tsdb", "tsdb"),
+            ("blocks_storage", "bucket_store", "tsdb-sync"),
+        ]
+
+        # The Mimir coordinator doesn't know the actual location of where the data
+        # will reside in the workers. The following loop updates the path of each key
+        # in the data_mapping list to match the actual common root data directory where
+        # the data will reside.
+        for key, subkey, folder in data_mapping:
+            # Ensure the key exists in the config dictionary
+            config.setdefault(key, {})
+
+            # Check if the subkey exists in the corresponding key's configuration
+            if subkey in config[key]:
+                # Update the subkey based on its type
+                # Both "data_dir" and "rule_path" in Mimir config don't have subkeys
+                if "data_dir" == subkey or subkey == "rule_path":
+                    config[key][subkey] = str(self._root_data_dir / folder)
```

Comment:
Would it be possible simplify with something like:
```python
if alertmanager := config.get("alertmanager"):
    alertmanager["data_dir"] = str(self._root_data_dir / "data-alertmanager")

if blocks_storage := config.get("blocks_storage"):
    blocks_storage.update({
        "filesystem": "blocks",
        "tsdb": "tsdb",
        "bucket_store": "bucket_store"
    })
```

Reply:
I feel like this is will indeed simplify the logic yet will still have the same drawbacks of having to update the path in the worker whenever a new field is added to the config, I pursued your value templating recommendation and came up with the below and was wondering if you favor one over the other or maybe if none of them is actually better than the original approach

1. Using the Template Python lib

```
def replace_path(config, root_data_dir="/newfield/data"):
    for key, value in config.items():
        if isinstance(value, dict):
            replace_path(value, root_data_dir)
        elif isinstance(value, Template):
            config[key] = value.substitute(root_data_dir=root_data_dir)

new_cfg = {
    'common': {},
    'alertmanager': {'data_dir': Template('$root_data_dir/data-alertmanager')},
    'alertmanager_storage': {'filesystem': {'dir': Template('$root_data_dir/data-alertmanager-recovery')}},
    'compactor': {'data_dir': Template('$root_data_dir/data-compactor')},
    'ruler': {'rule_path': Template('$root_data_dir/data-ruler')},
    'ruler_storage': {'filesystem': {'dir': Template('$root_data_dir/rules')}},
    'blocks_storage': {
        'bucket_store': {'sync_dir': Template('$root_data_dir/tsdb-sync')},
        'filesystem': {'dir': Template('$root_data_dir/blocks')},
        'tsdb': {'dir': Template('$root_data_dir/tsdb')},
    }
}
replace_path(new_cfg)
```
Recursively iterate over the config and substitute each Template str with the new path, The down side I see to this is having to initialize a template for each config due to the fact that `TypeError: unsupported operand type(s) for +: 'Template' and 'str'`, hence the need to do smth like below

```
alertmanager_data_dir_template = Template("$root_data_dir/data-alertmanager")
alertmanager_recovery_dir_template = Template("$root_data_dir/data-alertmanager-recovery")
compactor_data_dir_template = Template("$root_data_dir/data-compactor")
ruler_data_dir_template = Template("$root_data_dir/data-ruler")
ruler_storage_dir_template = Template("$root_data_dir/rules")
tsdb_sync_dir_template = Template("$root_data_dir/tsdb-sync")
blocks_dir_template = Template("$root_data_dir/blocks")
tsdb_dir_template = Template("$root_data_dir/tsdb")
```
1. Using placeholders

```
def replace_path(config, placeholder, new_path):
    for key, value in config.items():
        if isinstance(value, dict):
            replace_path(value, placeholder, new_path)
        elif isinstance(value, str):
            config[key] = value.format(**{placeholder: new_path})

new_cfg = {
    "common": {},
    "alertmanager": {"data_dir": "{path_placeholder}/data-alertmanager"},
    "alertmanager_storage": {"filesystem": {"dir": "{path_placeholder}/data-alertmanager"}},
    "compactor": {"data_dir": "{path_placeholder}/data-alertmanager"},
    "ruler": {"rule_path": "{path_placeholder}/data-alertmanager"},
    "ruler_storage": {"filesystem": {"dir": "{path_placeholder}/data-alertmanager"}},
    "blocks_storage": {
        "bucket_store": {"sync_dir": "{path_placeholder}/data-alertmanager"},
        "filesystem": {"dir": "{path_placeholder}/data-alertmanager"},
        "tsdb": {"dir": "{path_placeholder}/data-alertmanager"},
    },
}

replace_path(new_cfg, "path_placeholder", "/new/path")
```

This to me feels better due to the fact that this "placeholder" can still be `_root_data_dir = Path("{path_placeholder}")` and this will be the only code change needed in the coordinator, yet it might have some draw backs that I can't really see, @sed-i  Any thoughts??

Reply:
Disregarded the above in favor of using hard-coded path in the coordinator that is identical to the one defined in the worker's metadata